File: MIDTERM-LAB-3/MP05_ExtractColumn.js
/**
 * =====================================================
 * Student Name    : Alidio, Jacob A.
 * Course          : Programming 2
 * Machine Problem : MP05 – Extract Values from a Column
 * School          : University of Perpetual Help System DALTA, Molino Campus
 * Date            : April 8, 2026
 * Runtime         : Node.js  (run with: node MP05_ExtractColumn.js)
 *
 * Description:
 *   Prompts the user for a CSV file path and a column name. The
 *   program reads the file with Node.js fs, locates the column in
 *   the header, then prints every value in that column row by row.
 *   Row numbers accompany each value for easy reference. Empty
 *   fields are labelled "(empty)" instead of printing blank lines.
 * =====================================================
 */

// ── SECTION 1: Module Imports ────────────────────────────────────────
const fs       = require('fs');       // fs reads the CSV from disk
const readline = require('readline'); // readline lets us prompt the user

// ── SECTION 2: User Input ────────────────────────────────────────────
// rl connects the terminal's input/output so we can ask questions
const rl = readline.createInterface({
    input : process.stdin,
    output: process.stdout
});

const line = '='.repeat(52);

console.log(line);
console.log('  MP05 – Extract Values from a Column');
console.log(line);

// Ask for the file path first, then ask for the column name inside the callback
rl.question('  Enter CSV file path  : ', (filePath) => {
    rl.question('  Enter column name    : ', (targetCol) => {
        rl.close();
        extractColumn(filePath.trim(), targetCol.trim());
    });
});

// ── SECTION 3: Column Extraction Function ───────────────────────────
/**
 * Reads the CSV, finds the target column by name, and prints its value
 * for every data row. Handles missing columns and read errors gracefully.
 *
 * @param {string} filePath   - path to the CSV file
 * @param {string} targetCol  - name of the column to extract
 */
function extractColumn(filePath, targetCol) {

    // ── Step 1: Read the file ──
    let rawContent;
    try {
        rawContent = fs.readFileSync(filePath, 'utf8');
    } catch (err) {
        console.log('[ERROR] Could not read file: ' + err.message);
        return;
    }

    // ── Step 2: Split into non-empty lines ──
    const lines = rawContent
        .split('\n')
        .map(l => l.trim())
        .filter(l => l.length > 0);

    if (lines.length === 0) {
        console.log('[ERROR] File is empty.');
        return;
    }

    // ── Step 3: Locate the column index in the header ──
    // headers is an array of column name strings from the first line
    const headers = lines[0].split(',').map(h => h.trim());

    // colIndex will hold the 0-based position of the target column
    // findIndex returns -1 if no match is found
    const colIndex = headers.findIndex(
        h => h.toLowerCase() === targetCol.toLowerCase()
    );

    if (colIndex === -1) {
        console.log(`[ERROR] Column "${targetCol}" not found.`);
        console.log('  Available columns:');
        headers.forEach(h => console.log(`    - ${h}`));
        return;
    }

    // ── Step 4: Print column values from each data row ──
    console.log();
    console.log(line);
    console.log(`  Column : ${headers[colIndex]}`);
    console.log(line);
    console.log(`  ${'Row'.padEnd(6)}  Value`);
    console.log('  ' + '-'.repeat(48));

    let extracted = 0;  // counts how many values were printed

    // Data rows start at index 1 (index 0 is the header)
    for (let i = 1; i < lines.length; i++) {

        const fields = lines[i].split(',');  // split row into individual field values
        const rowNum = i;                    // 1-based row number for the user

        // Safely access the column; fall back to empty string if row is short
        const value   = colIndex < fields.length ? fields[colIndex].trim() : '';
        const display = value.length === 0 ? '(empty)' : value;

        console.log(`  ${String(rowNum).padEnd(6)}  ${display}`);
        extracted++;
    }

    // ── Step 5: Summary footer ──
    console.log(line);
    console.log(`  Total values extracted : ${extracted}`);
    console.log(line);
}

File: MIDTERM-LAB-3/MP05_ExtractColumn.java
/**
 * =====================================================
 * Student Name    : Alidio, Jacob A.
 * Course          : Programming 2
 * Machine Problem : MP05 – Extract Values from a Column
 * School          : University of Perpetual Help System DALTA, Molino Campus
 * Date            : April 8, 2026
 * Runtime         : Java (compile: javac MP05_ExtractColumn.java)
 *                           (run:     java  MP05_ExtractColumn)
 *
 * Description:
 *   Prompts the user for a CSV file path and a column name. The
 *   program locates that column in the header row, then extracts
 *   and prints every value found in that column across all data
 *   rows. Row numbers are shown alongside each value for easy
 *   reference. Empty values are flagged as "(empty)".
 * =====================================================
 */

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.Scanner;

public class MP05_ExtractColumn {

    // ── SECTION 1: Entry Point ───────────────────────────────────────────
    public static void main(String[] args) {

        // Scanner captures both the file path and column name from the user
        Scanner scanner = new Scanner(System.in);

        System.out.println("=".repeat(52));
        System.out.println("  MP05 – Extract Values from a Column");
        System.out.println("=".repeat(52));
        System.out.print("  Enter CSV file path  : ");
        String filePath  = scanner.nextLine().trim();   // path to the CSV

        System.out.print("  Enter column name    : ");
        String targetCol = scanner.nextLine().trim();   // column the user wants
        scanner.close();

        extractColumn(filePath, targetCol);
    }

    // ── SECTION 2: Column Extraction ────────────────────────────────────
    /**
     * Finds the target column by name in the header, then prints every
     * value from that column in each data row.
     *
     * @param filePath   path to the CSV file
     * @param targetCol  the column name to extract (case-insensitive match)
     */
    static void extractColumn(String filePath, String targetCol) {

        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {

            // ── Step 1: Read and parse the header line ──
            String headerLine = br.readLine();
            if (headerLine == null) {
                System.out.println("[ERROR] File is empty.");
                return;
            }

            // headers array holds each column name from the first CSV line
            String[] headers = headerLine.split(",", -1);

            // colIndex is the position (0-based) of the target column; -1 = not found
            int colIndex = -1;
            for (int i = 0; i < headers.length; i++) {
                // Case-insensitive trim comparison so "Name" matches "name"
                if (headers[i].trim().equalsIgnoreCase(targetCol)) {
                    colIndex = i;
                    break;
                }
            }

            // ── Step 2: Abort if column not found ──
            if (colIndex == -1) {
                System.out.println("[ERROR] Column \"" + targetCol + "\" not found.");
                System.out.println("  Available columns:");
                for (String h : headers) {
                    System.out.println("    - " + h.trim());
                }
                return;
            }

            // ── Step 3: Print header for the results table ──
            System.out.println();
            System.out.println("=".repeat(52));
            System.out.printf("  Column : %s%n", headers[colIndex].trim());
            System.out.println("=".repeat(52));
            System.out.printf("  %-6s  %s%n", "Row", "Value");
            System.out.println("  " + "-".repeat(48));

            // ── Step 4: Read data rows and print the target column value ──
            String line;        // current data line
            int rowNum = 1;     // 1-based row counter shown to the user
            int extracted = 0;  // count of values successfully extracted

            while ((line = br.readLine()) != null) {

                if (line.trim().isEmpty()) continue;  // skip blank lines

                // Split each data row by comma; -1 preserves trailing empties
                String[] fields = line.split(",", -1);

                // Safely retrieve the value; guard against short rows
                String value = (colIndex < fields.length)
                               ? fields[colIndex].trim()
                               : "";

                // Display "(empty)" when the field has no content
                String display = value.isEmpty() ? "(empty)" : value;

                System.out.printf("  %-6d  %s%n", rowNum, display);
                rowNum++;
                extracted++;
            }

            // ── Step 5: Summary footer ──
            System.out.println("=".repeat(52));
            System.out.printf("  Total values extracted : %d%n", extracted);
            System.out.println("=".repeat(52));

        } catch (IOException e) {
            System.out.println("[ERROR] Could not read file: " + e.getMessage());
        }
    }
}